src/main/scala/firrtl/passes/ExpandWhens.scala

```diff
@@ -0,0 +1,201 @@
+/*
+Copyright (c) 2014 - 2016 The Regents of the University of
+California (Regents). All Rights Reserved.  Redistribution and use in
+source and binary forms, with or without modification, are permitted
+provided that the following conditions are met:
+   * Redistributions of source code must retain the above
+     copyright notice, this list of conditions and the following
+     two paragraphs of disclaimer.
+   * Redistributions in binary form must reproduce the above
+     copyright notice, this list of conditions and the following
+     two paragraphs of disclaimer in the documentation and/or other materials
+     provided with the distribution.
+   * Neither the name of the Regents nor the names of its contributors
+     may be used to endorse or promote products derived from this
+     software without specific prior written permission.
+IN NO EVENT SHALL REGENTS BE LIABLE TO ANY PARTY FOR DIRECT, INDIRECT,
+SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES, INCLUDING LOST PROFITS,
+ARISING OUT OF THE USE OF THIS SOFTWARE AND ITS DOCUMENTATION, EVEN IF
+REGENTS HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+REGENTS SPECIFICALLY DISCLAIMS ANY WARRANTIES, INCLUDING, BUT NOT
+LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+A PARTICULAR PURPOSE. THE SOFTWARE AND ACCOMPANYING DOCUMENTATION, IF
+ANY, PROVIDED HEREUNDER IS PROVIDED "AS IS". REGENTS HAS NO OBLIGATION
+TO PROVIDE MAINTENANCE, SUPPORT, UPDATES, ENHANCEMENTS, OR
+MODIFICATIONS.
+*/
+
+package firrtl.passes
+
+import firrtl._
+import firrtl.Utils._
+import firrtl.Mappers._
+import firrtl.PrimOps._
+import firrtl.WrappedExpression._
+
+// Datastructures
+import scala.collection.mutable.HashMap
+import scala.collection.mutable.LinkedHashMap
+import scala.collection.mutable.ArrayBuffer
+
+import annotation.tailrec
+
+/** Expand Whens
+*
+* @note This pass does three things: remove last connect semantics,
+* remove conditional blocks, and eliminate concept of scoping.
+* @note Assumes bulk connects and isInvalids have been expanded
+* @note Assumes all references are declared
+*/
+object ExpandWhens extends Pass {
+  def name = "Expand Whens"
+
+  // ========== Expand When Utilz ==========
+  private def getEntries(
+      hash: LinkedHashMap[WrappedExpression, Expression],
+      exps: Seq[Expression]): LinkedHashMap[WrappedExpression, Expression] = {
+    val hashx = LinkedHashMap[WrappedExpression, Expression]()
+    exps foreach (e => if (hash.contains(e)) hashx(e) = hash(e))
+    hashx
+  }
+  private def getFemaleRefs(n: String, t: Type, g: Gender): Seq[Expression] = {
+    def getGender(t: Type, i: Int, g: Gender): Gender = times(g, get_flip(t, i, DEFAULT))
+    val exps = create_exps(WRef(n, t, ExpKind(), g))
+    val expsx = ArrayBuffer[Expression]()
+    for (i <- 0 until exps.size) {
```

Comment:
suggest not reusing variable name _i_ here -- the indices floating around are confusing enough as it is